File: test/language/array_types/java/array_types/PackedVariableArrayStructRecursionTest.java
package array_types;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.File;

import zserio.runtime.ZserioError;
import zserio.runtime.io.FileBitStreamReader;
import zserio.runtime.io.FileBitStreamWriter;
import array_types.packed_variable_array_struct_recursion.Block;
import array_types.packed_variable_array_struct_recursion.PackedVariableArray;

public class PackedVariableArrayStructRecursionTest
{
    @Test
    public void bitSizeOfLength1() throws IOException, ZserioError
    {
        checkBitSizeOf(VARIABLE_ARRAY_LENGTH1);
    }

    @Test
    public void bitSizeOfLength2() throws IOException, ZserioError
    {
        checkBitSizeOf(VARIABLE_ARRAY_LENGTH2);
    }

    @Test
    public void bitSizeOfLength3() throws IOException, ZserioError
    {
        checkBitSizeOf(VARIABLE_ARRAY_LENGTH3);
    }

    @Test
    public void writeReadLength1() throws IOException, ZserioError
    {
        checkWriteRead(VARIABLE_ARRAY_LENGTH1);
    }

    @Test
    public void writeReadLength2() throws IOException, ZserioError
    {
        checkWriteRead(VARIABLE_ARRAY_LENGTH2);
    }

    @Test
    public void writeReadLength3() throws IOException, ZserioError
    {
        checkWriteRead(VARIABLE_ARRAY_LENGTH3);
    }

    private void checkBitSizeOf(int numElements) throws IOException, ZserioError
    {
        final PackedVariableArray packedVariableArray = createPackedVariableArray(numElements);

        final int unpackedBitsizeOf = calcUnpackedVariableArrayBitSize(numElements);
        final int packedBitsizeOf = packedVariableArray.bitSizeOf();

        final double minCompressionRatio = 0.9;
        assertTrue(unpackedBitsizeOf * minCompressionRatio > packedBitsizeOf, () ->
                "Unpacked array has " + unpackedBitsizeOf + " bits, packed array has " + packedBitsizeOf +
                " bits, " + "compression ratio is " + packedBitsizeOf * 100.0 / unpackedBitsizeOf + "%!");
    }

    private void checkWriteRead(int numElements) throws IOException, ZserioError
    {
        final PackedVariableArray packedVariableArray = createPackedVariableArray(numElements);
        final File file = new File(BLOB_NAME_BASE + numElements + ".blob");
        final FileBitStreamWriter writer = new FileBitStreamWriter(file);
        packedVariableArray.write(writer);
        writer.close();

        assertEquals(packedVariableArray.bitSizeOf(), writer.getBitPosition());
        assertEquals(packedVariableArray.initializeOffsets(), writer.getBitPosition());

        final FileBitStreamReader reader = new FileBitStreamReader(file);
        final PackedVariableArray readPackedVariableArray = new PackedVariableArray(reader);
        assertEquals(packedVariableArray, readPackedVariableArray);
    }

    private PackedVariableArray createPackedVariableArray(int numElements)
    {
        final short byteCount = 1;
        final Block[] blocks = new Block[numElements];
        for (int i = 0; i < numElements; ++i)
            blocks[i] = createBlock(byteCount, false);

        return new PackedVariableArray(byteCount, numElements, blocks);
    }

    private Block createBlock(short byteCount, boolean isLast)
    {
        final short[] dataBytes = new short[byteCount];
        for (short i = 0; i < byteCount; ++i)
            dataBytes[i] = i;
        if (isLast)
            return new Block(byteCount, dataBytes, (short)0, null);

        final short blockTerminator = (short)(byteCount + 1);
        final Block block = createBlock(blockTerminator, blockTerminator > 5);

        return new Block(byteCount, dataBytes, blockTerminator, block);
    }

    private int calcUnpackedVariableArrayBitSize(int numElements)
    {
        int bitSize = 8; // byteCount
        bitSize += 8; // numElements
        final short byteCount = 1;
        for (int i = 0; i < numElements; ++i)
            bitSize += calcUnpackedBlockBitSize(byteCount, false);

        return bitSize;
    }

    private int calcUnpackedBlockBitSize(short byteCount, boolean isLast)
    {
        int bitSize = 8 * byteCount; // dataBytes[byteCount]
        bitSize += 8; // blockTerminator
        if (!isLast)
        {
            final short blockTerminator = (short)(byteCount + 1);
            bitSize += calcUnpackedBlockBitSize(blockTerminator, blockTerminator > 5);
        }

        return bitSize;
    }

    private static final String BLOB_NAME_BASE = "packed_variable_array_struct_recursion_";

    private static final int VARIABLE_ARRAY_LENGTH1 = 100;
    private static final int VARIABLE_ARRAY_LENGTH2 = 500;
    private static final int VARIABLE_ARRAY_LENGTH3 = 1000;
}

File: test/language/array_types/cpp/PackedVariableArrayStructRecursionTest.cpp
#include "array_types/packed_variable_array_struct_recursion/PackedVariableArray.h"
#include "gtest/gtest.h"
#include "zserio/SerializeUtil.h"

namespace array_types
{
namespace packed_variable_array_struct_recursion
{

class PackedVariableArrayStructRecursionTest : public ::testing::Test
{
protected:
    void fillPackedVariableArray(PackedVariableArray& packedVariableArray, size_t numElements)
    {
        const uint8_t byteCount = 1;
        packedVariableArray.setByteCount(byteCount);
        packedVariableArray.setNumElements(static_cast<uint32_t>(numElements));
        auto& blocks = packedVariableArray.getPackedBlocks();
        blocks.reserve(numElements);
        for (size_t i = 0; i < numElements; ++i)
        {
            blocks.push_back(createBlock(byteCount, false));
        }
        packedVariableArray.initializeChildren();
        packedVariableArray.initializeOffsets();
    }

    Block createBlock(uint8_t byteCount, bool isLast)
    {
        Block block;
        auto& dataBytes = block.getDataBytes();
        dataBytes.reserve(byteCount);
        for (uint8_t i = 0; i < byteCount; ++i)
        {
            dataBytes.push_back(i);
        }

        if (isLast)
        {
            block.setBlockTerminator(0);
            return block;
        }
        else
        {
            const uint8_t blockTerminator = static_cast<uint8_t>(byteCount + 1);
            block.setBlockTerminator(blockTerminator);
            block.setNextData(createBlock(blockTerminator, blockTerminator > 5));
            return block;
        }
    }

    size_t getUnpackedVariableArrayBitSize(size_t numElements)
    {
        size_t bitSize = 8; // byteCount
        bitSize += 8; // numElements
        const uint8_t byteCount = 1;
        for (size_t i = 0; i < numElements; ++i)
        {
            bitSize += getUnpackedBlockBitSize(byteCount, false);
        }

        return bitSize;
    }

    size_t getUnpackedBlockBitSize(uint8_t byteCount, bool isLast)
    {
        size_t bitSize = 8U * byteCount; // dataBytes[byteCount]
        bitSize += 8; // blockTerminator
        if (!isLast)
        {
            const uint8_t blockTerminator = static_cast<uint8_t>(byteCount + 1);
            bitSize += getUnpackedBlockBitSize(blockTerminator, blockTerminator > 5);
        }

        return bitSize;
    }

    void checkBitSizeOf(size_t numElements)
    {
        PackedVariableArray packedVariableArray;
        fillPackedVariableArray(packedVariableArray, numElements);

        const double unpackedBitSize = static_cast<double>(getUnpackedVariableArrayBitSize(numElements));
        const double packedBitSize = static_cast<double>(packedVariableArray.bitSizeOf());
        const double minCompressionRatio = 0.9;
        ASSERT_GT(unpackedBitSize * minCompressionRatio, packedBitSize)
                << "Unpacked array has " << std::to_string(unpackedBitSize) << " bits, "
                << "packed array has " << std::to_string(packedBitSize) << " bits, "
                << "compression ratio is " << std::to_string(packedBitSize / unpackedBitSize * 100) << "%!";
    }

    void checkWriteRead(size_t numElements)
    {
        PackedVariableArray packedVariableArray;
        fillPackedVariableArray(packedVariableArray, numElements);

        zserio::BitStreamWriter writer(bitBuffer);
        packedVariableArray.write(writer);

        ASSERT_EQ(packedVariableArray.bitSizeOf(), writer.getBitPosition());
        ASSERT_EQ(packedVariableArray.initializeOffsets(), writer.getBitPosition());

        zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
        PackedVariableArray readPackedVariableArray(reader);
        ASSERT_EQ(packedVariableArray, readPackedVariableArray);
    }

    void checkWriteReadFile(size_t numElements)
    {
        PackedVariableArray packedVariableArray;
        fillPackedVariableArray(packedVariableArray, numElements);

        const std::string fileName = BLOB_NAME_BASE + std::to_string(numElements) + ".blob";
        zserio::serializeToFile(packedVariableArray, fileName);

        PackedVariableArray readPackedVariableArray =
                zserio::deserializeFromFile<PackedVariableArray>(fileName);
        ASSERT_EQ(packedVariableArray, readPackedVariableArray);
    }

    static const std::string BLOB_NAME_BASE;

    static const size_t VARIABLE_ARRAY_LENGTH1;
    static const size_t VARIABLE_ARRAY_LENGTH2;
    static const size_t VARIABLE_ARRAY_LENGTH3;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(20 * 1024 * 8);
};

const std::string PackedVariableArrayStructRecursionTest::BLOB_NAME_BASE =
        "language/array_types/packed_variable_array_struct_recursion_";

const size_t PackedVariableArrayStructRecursionTest::VARIABLE_ARRAY_LENGTH1 = 100;
const size_t PackedVariableArrayStructRecursionTest::VARIABLE_ARRAY_LENGTH2 = 500;
const size_t PackedVariableArrayStructRecursionTest::VARIABLE_ARRAY_LENGTH3 = 1000;

TEST_F(PackedVariableArrayStructRecursionTest, bitSizeOfLength1)
{
    checkBitSizeOf(VARIABLE_ARRAY_LENGTH1);
}

TEST_F(PackedVariableArrayStructRecursionTest, bitSizeOfLength2)
{
    checkBitSizeOf(VARIABLE_ARRAY_LENGTH2);
}

TEST_F(PackedVariableArrayStructRecursionTest, bitSizeOfLength3)
{
    checkBitSizeOf(VARIABLE_ARRAY_LENGTH3);
}

TEST_F(PackedVariableArrayStructRecursionTest, writeReadLength1)
{
    checkWriteRead(VARIABLE_ARRAY_LENGTH1);
}

TEST_F(PackedVariableArrayStructRecursionTest, writeReadLength2)
{
    checkWriteRead(VARIABLE_ARRAY_LENGTH2);
}

TEST_F(PackedVariableArrayStructRecursionTest, writeReadLength3)
{
    checkWriteRead(VARIABLE_ARRAY_LENGTH3);
}

TEST_F(PackedVariableArrayStructRecursionTest, writeReadFileLength1)
{
    checkWriteReadFile(VARIABLE_ARRAY_LENGTH1);
}

TEST_F(PackedVariableArrayStructRecursionTest, writeReadFileLength2)
{
    checkWriteReadFile(VARIABLE_ARRAY_LENGTH2);
}

TEST_F(PackedVariableArrayStructRecursionTest, writeReadFileLength3)
{
    checkWriteReadFile(VARIABLE_ARRAY_LENGTH3);
}

} // namespace packed_variable_array_struct_recursion
} // namespace array_types
